pkg/etcd-backup-operator/controller/s3backup.go

```diff
@@ -0,0 +1,233 @@
+		pod := &podList.Items[i]
+		if pod.Status.Phase == v1.PodRunning {
+			pods = append(pods, pod)
+		}
+	}
+
+	if len(pods) == 0 {
+		msg := "no running etcd pods found"
+		logrus.Warning(msg)
+		return fmt.Errorf(msg)
+	}
+	member, rev := getMemberWithMaxRev(pods, nil)
+	if member == nil {
+		logrus.Warning("no reachable member")
+		return fmt.Errorf("no reachable member")
+	}
+
+	log.Printf("saving backup for cluster (%s)", b.clusterName)
+	if err := b.writeSnap(member, rev); err != nil {
+		err = fmt.Errorf("write snapshot failed: %v", err)
+		return err
+	}
+	return nil
+}
+
+func (b *s3Backup) writeSnap(m *etcdutil.Member, rev int64) error {
+	cfg := clientv3.Config{
+		Endpoints:   []string{m.ClientURL()},
+		DialTimeout: constants.DefaultDialTimeout,
+		TLS:         nil,
+	}
+	etcdcli, err := clientv3.New(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client (%v)", err)
+	}
+	defer etcdcli.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
+	cancel()
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
+	defer cancel()
+	rc, err := etcdcli.Maintenance.Snapshot(ctx)
+	defer rc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to receive snapshot (%v)", err)
+	}
+	return b.saveToS3(resp.Version, rev, rc)
+}
+
+// NewS3Uploader returns a Uploader that is configured with proper aws config and credentials derived from Kubernetes secret.
+func NewS3Uploader(bucket, prefix string, secret *v1.Secret) (*s3manager.Uploader, error) {
+	so, err := newAWSConfig(secret, prefix)
+	if err != nil {
+		return nil, err
+	}
+	return NewFromSessionOpt(bucket, prefix, *so)
+}
+
+func newAWSConfig(secret *v1.Secret, prefix string) (*session.Options, error) {
+	dir := filepath.Join(operatorRoot, "aws", prefix)
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, err
+	}
+
+	options := &session.Options{}
+	options.SharedConfigState = session.SharedConfigEnable
+
+	creds := secret.Data[api.AWSSecretCredentialsFileName]
+	if len(creds) != 0 {
+		credsFile := path.Join(dir, "credentials")
+		err := ioutil.WriteFile(credsFile, creds, 0600)
+		if err != nil {
+			return nil, fmt.Errorf("setup AWS config failed: write credentials file failed: %v", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, credsFile)
+	}
+
+	config := secret.Data[api.AWSSecretConfigFileName]
+	if config != nil {
+		configFile := path.Join(dir, "config")
+		err := ioutil.WriteFile(configFile, config, 0600)
+		if err != nil {
+			return nil, fmt.Errorf("setup AWS config failed: write config file failed: %v", err)
+		}
+		options.SharedConfigFiles = append(options.SharedConfigFiles, configFile)
+	}
+
+	return options, nil
+}
+
+func NewFromSessionOpt(bucket, prefix string, so session.Options) (*s3manager.Uploader, error) {
```

Comment:
This function should also be unexported and lowercased.

Reply:
fixed